server/command_handler.go

```diff
@@ -44,13 +50,122 @@ func subscribeHandler(server *Server, client *Client, frm *frame.Frame) {
 		sendError(client, nil, "destination field is required in the subscribe frame")
 		return
 	}
-	if _, ok := server.subscribers[destination]; !ok {
+	if _, ok := server.destinations[destination]; !ok {
 		sendError(client, nil, destination+"is not supported in the server")
 		return
 	}
-	server.subscribersLock.Lock()
-	subList := server.subscribers[destination]
-	newSubList := append(subList, client)
-	server.subscribers[destination] = newSubList
-	server.subscribersLock.Unlock()
+	newSubscriber := &subscriber{
+		client:       client,
+		isSubscribed: true,
+	}
+	server.destinationsLock.Lock()
+	subList := server.destinations[destination]
+	newSubList := append(subList, newSubscriber)
+	server.destinations[destination] = newSubList
+	server.destinationsLock.Unlock()
+}
+
+func unsubscribeHandler(server *Server, client *Client, frm *frame.Frame) {
+	destination, ok := frm.Headers["destination"]
+	if !ok {
+		sendError(client, nil, "destination field is required in the subscribe frame")
+		return
+	}
+	if _, ok := server.destinations[destination]; !ok {
+		sendError(client, nil, destination+"is not supported in the server")
+		return
+	}
+	subscribers := server.destinations[destination]
+	for _, sub := range subscribers {
+		if sub.client == client {
+			sub.isSubscribed = false
```

Comment:
BUG;
Compare by reference will not work. (note same error is in server.hasClient)

Reply:
Can you elaborate more?
new subscriber is being initialized with a client pointer in the  [subscribeHandler](https://github.com/MohamedShetewi/stomp-in-go/blob/c217d9e2c37f79bb9a5074947d3744cf9a3a7b55/server/command_handler.go#L42) func. And here I think we are passing the same client pointer passed in the `subscribeHandler` func.